src/methods/trend_based.jl

```diff
@@ -34,78 +48,73 @@ struct TFTDRandomFourier <: Surrogate
-    ldiv!(cholesky!(Symmetric([T(N) sum(x); zero(T) sum(abs2, x)], :U)), [sum(y), dot(x, y)])
-end
-
-function linear_trend(x)
-    l = linreg(0.0:1.0:length(x)-1.0 |> collect, x)
-    trendᵢ(xᵢ) = l[1] + l[2] * xᵢ
-    return trendᵢ.(x)
-end
+const TFTD2 = TFTDRandomFourier
 
 function surrogenerator(x::AbstractVector, rf::TFTDRandomFourier, rng = Random.default_rng())
     # Detrended time series
     m = mean(x)
     trend = linear_trend(x)
-
     x̂ = x .- m .- trend
 
-    # Pre-plan Fourier transforms
+    # Pre-plan and allocate Fourier transform
     forward = plan_rfft(x̂)
     inverse = plan_irfft(forward * x̂, length(x̂))
- 
-    # Pre-compute 𝓕
-    𝓕 = forward*x̂
- 
+    𝓕 = forward * x̂
+    n = length(𝓕)
+
     # Polar coordinate representation of the Fourier transform
     rx = abs.(𝓕)
     ϕx = angle.(𝓕)
-    n = length(𝓕)
-  
-    # These are updated during iteration procedure
-    𝓕new = Vector{Complex{Float64}}(undef, length(𝓕))
-    𝓕s = Vector{Complex{Float64}}(undef, length(𝓕))
-    ϕs = Vector{Complex{Float64}}(undef, length(𝓕))
+    ϕs = similar(ϕx)
+
+    permutation = zeros(Int, length(x))
+    idxs = collect(1:length(x))
+    
+    # Initialize surrogate
+    s = similar(x)
  
     init = (forward = forward, inverse = inverse,
         rx = rx, ϕx = ϕx, n = n, m = m,
-        𝓕new = 𝓕new, 𝓕s = 𝓕s, ϕs = ϕs, 
-        trend = trend, x̂ = x̂)
+        𝓕 = 𝓕, ϕs = ϕs, 
+        trend = trend, x̂ = x̂,
+        permutation, idxs)
 
-    return SurrogateGenerator(rf, x, init, rng)
+    return SurrogateGenerator(rf, x, s, init, rng)
 end
 
 function (sg::SurrogateGenerator{<:TFTDRandomFourier})()
     fϵ = sg.method.fϵ
+    s = sg.s
 
     init_fields = (:forward, :inverse,
         :rx, :ϕx, :n, :m,
-        :𝓕new, :𝓕s, :ϕs, :trend, :x̂)
+        :𝓕, :ϕs, :trend, :x̂, :permutation, :idxs)
 
     forward, inverse,
         rx, ϕx, n, m,
-        𝓕new, 𝓕s, ϕs, trend, x̂ = getfield.(Ref(sg.init), init_fields)
+        𝓕, ϕs, trend, x̂,
+        permutation, idxs = getfield.(Ref(sg.init), init_fields)
 
     # Surrogate starts out as a random permutation of x̂
-    s = x̂[StatsBase.sample(sg.rng, 1:length(x̂), length(x̂); replace = false)]
-    𝓕s .= forward*s
-    ϕs .= angle.(𝓕s)
+    sample!(sg.rng, idxs, permutation; replace = false)
+    s .= @view x̂[permutation]
+
+    mul!(𝓕, forward, s)
+    ϕs .= angle.(𝓕)
 
     # Frequencies are ordered from lowest when taking the Fourier
     # transform, so by keeping the 1:n_preserve first phases intact,
     # we are only randomizing the high-frequency components of the
     # signal.
     n_preserve = ceil(Int, abs(fϵ * n))
-    ϕs[1:n_preserve] .= ϕx[1:n_preserve]
+    ϕs[1:n_preserve] .= @view ϕx[1:n_preserve]
     
     # Updated spectrum is the old amplitudes with the mixed phases.
-    𝓕new .= rx .* exp.(ϕs .* 1im)
+    𝓕 .= rx .* exp.(ϕs .* 1im)
+
+    # Unfortunately, we can't do inverse transform in-place yet, but 
```

Comment:
```suggestion
    # TODO: Unfortunately, we can't do inverse transform in-place yet, but 
```

(just so we can find this again in the future)